src/providers/token-fee-provider.ts

```diff
@@ -0,0 +1,164 @@
+      .map((token) => token.address.toLowerCase())
+      .uniq()
+      .value();
+
+    const addresses: string[] = [];
+    const tokenToResult: { [tokenAddress: string]: TokenFeeResult } = {};
+
+    // Check if we have cached token validation results for any tokens.
+    for (const address of addressesRaw) {
+      const cachedValue = await this.tokenFeeCache.get(this.CACHE_KEY(this.chainId, address));
+      if (cachedValue) {
+        tokenToResult[address] = cachedValue
+      } else if (this.allowList.has(address)) {
+        tokenToResult[address] = DEFAULT_TOKEN_FEE_RESULT;
+      } else {
+        addresses.push(address);
+      }
+    }
+
+    log.info(
+      `Got token fee results for ${addressesRaw.length - addresses.length
+      } tokens from cache. Getting ${addresses.length} on-chain.`
+    );
+
+    if (addresses.length > 0) {
+      const functionParams = addresses
+        .map((address) => [address, this.BASE_TOKEN, this.amountToFlashBorrow]) as [string, string, string][];
+
+      // We use the validate function instead of batchValidate to avoid poison pill problem.
+      // One token that consumes too much gas could cause the entire batch to fail.
+      const multicallResult =
+      await this.multicall2Provider.callSameFunctionOnContractWithMultipleParams<
+      [string, string, string], // address, base token address, amount to borrow
+      [TokenFeeResult]
+      >({
+        address: this.tokenValidatorAddress,
+        contractInterface: TokenFeeDetector__factory.createInterface(),
+        functionName: 'validate',
+        functionParams: functionParams,
+        providerConfig,
+        additionalConfig: {
+          gasLimitPerCallOverride: this.gasLimitPerCall,
+        },
+      });
+
+      for (let i = 0; i < multicallResult.results.length; i++) {
+        const resultWrapper = multicallResult.results[i]!;
+        const tokenAddress = addresses[i]!;
+        const token = tokenAddressToToken[tokenAddress]!;
+
+        // Could happen if the tokens transfer consumes too much gas so we revert. Just
+        // drop the token in that case.
+        if (!resultWrapper.success || resultWrapper.result.length < 1) {
+          log.warn(
+            { result: resultWrapper },
+            `Failed to validate token ${token.symbol}`
+          );
+
+          continue;
+        }
+        if (resultWrapper.result.length > 1) {
+          log.warn(
+            { result: resultWrapper },
+            `Unexpected result length: ${resultWrapper.result.length}`
+          );
+        }
+        const validationResult = resultWrapper.result[0]!;
+
+        tokenToResult[tokenAddress] =
+          validationResult as TokenFeeResult;
+
+        await this.tokenFeeCache.set(
+          this.CACHE_KEY(this.chainId, tokenAddress),
+          tokenToResult[tokenAddress]!
+        );
+      }
+    }
+
+    return {
+      getFeesByToken: (token: Token) =>
+        tokenToResult[token.address.toLowerCase()],
+    };
+  }
```

Comment:
I'm also not a fan of returning anonymous functions as part of the response rather than objects of a specific class.

Could we instead do:
```
export class TokenFeeResults {
  constructor(private readonly tokenToResult: { [tokenAddress: string]: TokenFeeResult }) {}
  
  getFeesByToken(token: Token): TokenFeeResult | undefined {
    return this.tokenToResult[token.address.toLowerCase()];
  }
}
```

and here return:
```
  return new TokenFeeResults(tokenToResult);
```